src/Uno.UI/UI/Xaml/Data/ItemIndexRange.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using DirectUI;
+using DirectUI.Components;
+
+namespace Windows.UI.Xaml.Data;
+
+/// <summary>
+/// Provides info about a range of items in the data source.
+/// </summary>
+public partial class ItemIndexRange
+{
+	/// <summary>
+	/// Initializes an instance of the ItemIndexRange class.
+	/// </summary>
+	/// <param name="firstIndex">The index of the first item in the instance of the ItemIndexRange class.</param>
+	/// <param name="length">The number of items in the instance of the ItemIndexRange class.</param>
+	public ItemIndexRange(int firstIndex, uint length)
+	{
+		this.FirstIndex = firstIndex;
+		this.Length = length;
+	}
+
+	/// <summary>
+	/// Gets the index of the first item in the instance of the ItemIndexRange class.
+	/// </summary>
+	public int FirstIndex { get; }
+
+	/// <summary>
+	/// Gets the index of the last item in the instance of the ItemIndexRange class.
+	/// </summary>
+	public int LastIndex => FirstIndex + (int)Length - 1;
+
+	/// <summary>
+	/// Gets the number of items in the instance of the ItemIndexRange class.
+	/// </summary>
+	public uint Length { get; }
+
+	public override bool Equals(object obj)
+	{
+		return obj is ItemIndexRange other &&
+			other.FirstIndex == FirstIndex &&
+			other.Length == Length;
+	}
+
+	public override int GetHashCode() => (FirstIndex, Length).GetHashCode();
+}
+partial class ItemIndexRange
+{
```

Comment:
Why do you need to keep a partial here?

Reply:
probably not needed, i used to separate public implementation from microsoft source
the plan was to move it into a separate file `_Partial.cs`
ended up just porting a single method
i guess we should just merge the two partials?